pkg/controllers/nodeclass/status/nodeimage.go

```diff
@@ -0,0 +1,179 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/Azure/karpenter-provider-azure/pkg/apis/v1alpha2"
+	"github.com/Azure/karpenter-provider-azure/pkg/providers/imagefamily"
+	"github.com/awslabs/operatorpkg/reasonable"
+	"github.com/samber/lo"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/utils/pretty"
+)
+
+type NodeImageReconciler struct {
+	nodeImageProvider imagefamily.NodeImageProvider
+	cm                *pretty.ChangeMonitor
+}
+
+func NewNodeImageReconciler(provider imagefamily.NodeImageProvider) *NodeImageReconciler {
+	return &NodeImageReconciler{
+		nodeImageProvider: provider,
+		cm:                pretty.NewChangeMonitor(),
+	}
+}
+
+func (r *NodeImageReconciler) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("nodeclass.nodeimage").
+		For(&v1alpha2.AKSNodeClass{}).
+		WithOptions(controller.Options{
+			RateLimiter:             reasonable.RateLimiter(),
+			MaxConcurrentReconciles: 10,
+		}).
+		Complete(reconcile.AsReconciler(m.GetClient(), r))
+}
+
+// The image version reconciler will detect reasons to bump the node image version as follows in order:
+//
+// Scenario A: Update all image versions to latest
+//   - 1. Initializes the images versions for a newly created AKSNodeClass, based on customer configuration.
+//   - 2. Indirectly handle image bump for k8s upgrade
+//   - 3. Can indirectly handle bumps for any images unsupported by node features, if required to in the future
+//     Note: Currently there are no node features to be handled in this way.
+//   - 4. TODO: Update NodeImages to latest if in a MW (maintenance windows) [retrieved from ConfigMap]
+//
+// Scenario B: Calculate images to be updated based on delta of available images
+//   - 5. Handles update cases when customer changes image family, SIG usage, or other means of image selectors
+//   - 6. Handles softly adding newest image version of any newly supported SKUs by Karpenter
+func (r *NodeImageReconciler) Reconcile(ctx context.Context, nodeClass *v1alpha2.AKSNodeClass) (reconcile.Result, error) {
+	logger := logging.FromContext(ctx)
+	logger.Debug("nodeclass.nodeimage: starting reconcile")
+
+	nodeImages, err := r.nodeImageProvider.List(ctx, nodeClass)
+	if err != nil {
+		logger.Debug("nodeclass.nodeimage: err listing node images")
```

Comment:
Same comment as other file

Reply:
Also you don't even log the error here so this is doubly not useful as it is now?

Reply:
Removed, and cleaned up logging